maps/game_map.go

```diff
@@ -18,13 +18,68 @@ type GameMap interface {
 	UpdateBoard(previousBoardState *rules.BoardState, settings rules.Settings, editor Editor) error
 }
 
+// dimensions is an internal type used to encapsulate the configuration of a particular map size.
+type dimensions struct {
+	// Width is the width, in number of map squares, of the map.
+	// The value 0 has a special meaning to mean unlimited.
+	Width uint
+	// Height is the height, in number of map squares, of the map.
+	// The value 0 has a special meaning to mean unlimited.
+	Height uint
+}
+
+// sizes is an interal type used to encapsulate the configuration of supported map sizes.
+type sizes []dimensions
+
+// IsUnlimited reports whether the supported sizes are unlimited.
+// Note that even for unlimited sizes, there will be an upper bound that can actually be run and visualised.
+func (d sizes) IsUnlimited() bool {
+	return len(d) == 1 && d[0].Width == 0
+}
+
+// UnlimitedSizes creates sizes for a map that has no fixed sizes (supports unlimited sizes).
+func UnlimitedSizes() sizes {
+	return sizes{dimensions{Width: 0, Height: 0}}
+}
+
+// FixedSizes creates dimensions for a map that has 1 or more fixed sizes.
+// The arguments are expected to be multiples of 2, with the first of each pair being width
+// and the second of each pair being height.
+// Examples:
+// - FixedSizes(9,11) creates a map that supports only a width of 9 and a height of 11.
+// - FixedSizes(11,11,19,19) creates dimension for a map that supports sizes 11x11 and 19x19
+//
+// Panics if the list of sizes is not a multiple of 2.
+func FixedSizes(a, b uint, c ...uint) sizes {
```

Comment:
This is fancy and I like it :) But also I wonder if it's overly complex, vs. something like:
```
func FixedSizes(a ...dimension) sizes {
...
}
```

Using it like this doesn't feel that bad, and it might be better to be more explicit:
```
FixedSizes(dimension{11, 11}, dimension{19, 19}, dimension{7,7})
```

Reply:
Sounds good to me, but I'd suggest a slight change to this:

```
func FixedSizes(a dimension, b ...dimension) sizes {
...
}
```

The reason being it forces at least one and we don't have the error condition of having none. It's a pretty common Go design pattern that Iv'e seen when you want at least one thing when using the spread operator.

Reply:
Totally, makes sense 👍